content/advent-2017/messaging-framework.md

```diff
@@ -0,0 +1,709 @@
++++
+author = ["Orr Chen"]
+title = "Simple messaging framework using Go TCP server and Kafka"
+linktitle = "messaging framework"
+date = 2017-12-31T00:00:00Z
+series = ["Advent 2017"]
++++  
+  
+![System diagram](/static/postimages/advent-2017/messaging-framework/system-diagram.png)  
+I needed to create a simple framework to provide my endpoint devices ( doesn't matter which platform they run on ) the option to send and receive messages from my backend.  
+I require those messages to be managed by a message broker so that they can be processed in an asynchronous way.  
+The system contains 4 main layers, this article section is mainly about the first one:  
+1. **TCP servers** - Needs to keep maintain as many as possible concurrent tcp sockets with the endpoints. all of the endpoints messages will be processed on a different layer through the message broker. This keeps the TCP servers layer very thin and effective and to keep as many concurrent connection as possible, and Go is a good pick for it ( see [this article](https://medium.freecodecamp.org/million-websockets-and-go-cc58418460bb))  
```

Comment:
s/to keep maintain as many as possible concurrent tcp sockets with the endpoints/to maintain as many TCP sockets in synch with the endpoints as possible (more options work here)

s/all/All

s/layer through the/layer by the

s/This keeps the TCP servers layer very thin and effective and to keep as many concurrent connection as possible, and Go is a good pick for it/This keeps the TCP servers layer very thin and effective. I also want to keep as many concurrent connection as possible, and Go is a good choice for this